htlcswitch/switch_test.go

```diff
@@ -2690,3 +2692,407 @@ func TestInvalidFailure(t *testing.T) {
+				// to forward the htlc onwards, and reports on any
+				// forwarding errors rather than the success of the
+				// payment as a whole.
+				sendPayment, pid, _ := addThreeHopPayment(t, n)
+				if err := sendPayment(); err != nil {
+					t.Fatalf("could not send payment")
+				}
+
+				// Get the updates channel for each htlc event
+				// subscription and check that it contains the
+				// events the test expects.
+				test.checkEvents(
+					t, aliceEvents.Updates(),
+					bobEvents.Updates(),
+					carolEvents.Updates(),
+					channels,
+					// We expect htlc ids to increment with
+					// each iteration.
+					uint64(i),
+					pid,
+				)
+			}
+		})
+	}
+}
+
+// addThreeHopPayment is a helper function which returns a function which can
+// be used to send a payment over Alice -> Bob -> Carol in a three hop network.
+// NOTE: this function
+func addThreeHopPayment(t *testing.T,
+	n *threeHopNetwork) (func() error, uint64, lntypes.Hash) {
+
+	amount := lnwire.NewMSatFromSatoshis(btcutil.SatoshiPerBitcoin)
+
+	htlcAmt, totalTimelock, hops := generateHops(amount, testStartingHeight,
+		n.firstBobChannelLink, n.carolChannelLink)
+	blob, err := generateRoute(hops...)
+	if err != nil {
+		t.Fatal(err)
+	}
+	invoice, htlc, pid, err := generatePayment(
+		amount, htlcAmt, totalTimelock, blob,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = n.carolServer.registry.AddInvoice(*invoice, htlc.PaymentHash)
+	if err != nil {
+		t.Fatalf("unable to add invoice in carol registry: %v", err)
+	}
+
+	// With the invoice now added to Carol's registry, we create a function
+	// which can be used to send the payment from Alice -> Bob -> Carol.
+	sendFunc := func() error {
+		return n.aliceServer.htlcSwitch.SendHTLC(
+			n.firstBobChannelLink.ShortChanID(), pid, htlc,
+		)
+	}
+
+	return sendFunc, pid, htlc.PaymentHash
+}
+
+// getCircuits returns circuits for a send from Alice -> Bob -> Carol with
+// the htlc ID provided.
+func getCircuits(channels *clusterChannels,
+	htlcID uint64) (CircuitKey, CircuitKey, CircuitKey, CircuitKey) {
+
+	// Alice -> Bob, from alice's switch's perspective.
+	aliceToBob := CircuitKey{
+		ChanID: channels.aliceToBob.ShortChanID(),
+		HtlcID: htlcID,
+	}
+
+	// Alice -> Bob, from bob's switch's perspective.
+	bobFromAlice := CircuitKey{
+		ChanID: channels.bobToAlice.ShortChanID(),
+		HtlcID: htlcID,
+	}
+
+	// Bob -> Carol, from bob's switch's perspective.
+	bobToCarol := CircuitKey{
+		ChanID: channels.bobToCarol.ShortChanID(),
+		HtlcID: htlcID,
+	}
+
+	// Bob -> Carol, from carol's switch's perspective.
+	carolFromBob := CircuitKey{
+		ChanID: channels.carolToBob.ShortChanID(),
+		HtlcID: htlcID,
+	}
+
+	return aliceToBob, bobFromAlice, bobToCarol, carolFromBob
+}
+
+// checkSuccessful is a function that checks that Alice, Bob and Carol
+// have the correct set of event for a payment that is successfully
+// routed over Alice -> Bob -> Carol.
+func checkSuccessful(t *testing.T, alice, bob, carol <-chan interface{},
```

Comment:
There is quite a bit of assertion code here. Can't this test be written by just having three (one for every node) lists of fully-specified events that are expected? The actual events are then just read from the channels and deep-equaled with the expectations.

Reply:
Switched over to a set of results, still quite a lot of setup code, but I think it's better. Less duplication at least.